net/libfab/libfab.c

```diff
@@ -0,0 +1,3451 @@
+	struct epoll_event        ev;
+	int                       ev_cnt;
+	int                       ret;
+
+	/* Wait for the transfer machine to get started */
+	while (tm->ftm_state != FAB_TM_STARTED)
+		/* No-op */;
+	
+	while (tm->ftm_state != FAB_TM_SHUTDOWN) {
+		sched_yield();
+		ev_cnt = epoll_wait(tm->ftm_epfd, &ev, 1, FAB_WAIT_FD_TMOUT);
+
+		while (1) {
+			m0_mutex_lock(&tm->ftm_endlock);
+			if (tm->ftm_state == FAB_TM_SHUTDOWN) {
+				m0_mutex_unlock(&tm->ftm_endlock);
+				return;
+			}
+			
+			ret = libfab_tm_trylock(tm);
+			m0_mutex_unlock(&tm->ftm_endlock);
+			if (ret == 0) {
+				/* Got tm lock. 
+				   Let's continue processing events. */
+				break;
+			}
+		}
+		
+		M0_ASSERT(libfab_tm_is_locked(tm) && libfab_tm_invariant(tm));
+
+		/* Check the common queue of the transfer machine for events */
+		libfab_handle_connect_request_events(tm);
+		libfab_txep_comp_read(tm->ftm_tx_cq, tm);
+
+		if (ev_cnt > 0) {
+			ctx = ev.data.ptr;
+			if (ctx->evctx_type != FAB_COMMON_Q_EVENT) {
+				/* Check the private queue of the
+				   endpoint for events */
+				xep = ctx->evctx_ep;
+				aep = libfab_aep_get(xep);
+				libfab_txep_event_check(xep, aep, tm);
+				cq = aep->aep_rx_res.frr_cq;
+				libfab_rxep_comp_read(cq, xep, tm);
+			}
+		}
+
+		libfab_bufq_process(tm);
+		if (m0_time_sub(m0_time_now(), tm->ftm_tmout_check) >=
+								 m0_time(1,0))
+			libfab_tm_buf_timeout(tm);
+		libfab_tm_buf_done(tm);
+
+		M0_ASSERT(libfab_tm_invariant(tm));
+		libfab_tm_unlock(tm);
+	}
+}
+
+/** 
+ * Converts network end-point to its libfabric structure.
+ */
+static inline struct m0_fab__ep *libfab_ep(struct m0_net_end_point *net)
+{
+	return container_of(net, struct m0_fab__ep, fep_nep);
+}
+
+/**
+ * Compares the endpoint name with the passed name string and
+ * returns true if equal, or else returns false
+ * If name is null then the ipaddress and port fields are compared to check for
+ * a matching endpoint.
+ */
+static bool libfab_ep_cmp(struct m0_fab__ep *ep, const char *name,
+			  uint64_t *ep_name_n)
+{
+	bool ret = false;
+	M0_PRE(ep != NULL);
+	if (name == NULL) {
+		if (*ep_name_n == ep->fep_name_n)
+			ret = true;
+	} else {
+		if (strcmp(ep->fep_name_p.fen_str_addr, name) == 0)
+			ret = true;
+	}
+	return ret;
+}
+
+/**
+ * Search for the ep in the existing ep list using one of the following -
+ *   1) Name in str format      OR
+ *   2) ipaddr and port 
+ * If found then return the ep structure, or else create a new endpoint
+ * with the name
+ * Returns 0 if the endpoint is found/created succesfully,
+ * else returns error code.
+ */
+static int libfab_ep_find(struct m0_net_transfer_mc *tm, const char *name,
+			  struct m0_fab__ep_name *epn,
+			  struct m0_net_end_point **epp)
+{
+	struct m0_net_end_point  *net;
+	struct m0_fab__ep        *ep;
+	struct m0_fab__active_ep *aep;
+	struct m0_fab__tm        *ma;
+	uint64_t                  ep_name_n = 0;
+	char                      ep_str[LIBFAB_ADDR_STRLEN_MAX + 9] = {'\0'};
```

Comment:
The memset can be moved to line 1054 for optimization

Reply:
Done